qutip/core/cy/coefficient.pyx

```diff
@@ -365,167 +373,180 @@ def coeff(t, args):
 
     cpdef Coefficient copy(self):
         """Return a copy of the :obj:`Coefficient`."""
-        return InterCoefficient(self.coeff_np, self.tlist_np,
-                                self.second_np, self.constant)
+        return InterCoefficient.restore(*self.np_arrays, self.dt)
 
 
 cdef Coefficient add_inter(InterCoefficient left, InterCoefficient right):
-    if np.array_equal(left.tlist_np, right.tlist_np):
-        return InterCoefficient(left.coeff_np + right.coeff_np,
-                                left.tlist_np,
-                                left.second_np + right.second_np,
-                                left.constant
-                               )
+    """ Add two array coefficient with matching tlist into one."""
+    if (
+        left.np_arrays[0].shape == right.np_arrays[0].shape
+        and np.allclose(left.np_arrays[0], right.np_arrays[0],
+                        rtol=1e-15, atol=1e-15)
+        and (left.order == right.order)
+    ):
+        return InterCoefficient.restore(
+            left.np_arrays[0], left.np_arrays[1] + right.np_arrays[1],
+            left.dt
+        )
     else:
-        return SumCoefficient(left.copy(), right.copy())
-
-
-cdef class StepCoefficient(Coefficient):
-    """
-    A step function :obj:`Coefficient` whose values are specified in a numpy array.
-
-    At each point in time, the value of the coefficient is the most recent previous value given in the numpy array.
-
-    Parameters
-    ----------
-    coeff_arr : np.ndarray
-        The array of coefficient values to interpolate.
-
-    tlist : np.ndarray
-        An array of times corresponding to each coefficient value. The times must be
-        increasing, but do not need to be uniformly spaced.
-    """
-    cdef int n_t, constant
-    cdef double dt
-    cdef double[::1] tlist
-    cdef complex[::1] coeff_arr
-    cdef object tlist_np, coeff_np
-
-    def __init__(self, coeff_arr, tlist, _constant=None):
-        self.tlist_np = tlist
-        self.tlist = self.tlist_np
-        self.coeff_np = coeff_arr
-        self.coeff_arr = self.coeff_np
-        if _constant is None:
-            self.constant = np.allclose(np.diff(tlist), tlist[1]-tlist[0])
-        else:
-            self.constant = _constant
-        self.dt = tlist[1] - tlist[0]
-        self.args = {}
-
-    @cython.initializedcheck(False)
-    cdef complex _call(self, double t) except *:
-        cdef complex coeff
-        if self.constant:
-            coeff = _step_complex_cte(t, self.tlist, self.coeff_arr)
-        else:
-            coeff = _step_complex_t(t, self.tlist, self.coeff_arr)
-        return coeff
-
-    def __reduce__(self):
-        return (StepCoefficient, (self.coeff_np, self.tlist_np, self.constant))
-
-    cpdef Coefficient copy(self):
-        """Return a copy of the :obj:`Coefficient`."""
-        return StepCoefficient(self.coeff_np, self.tlist_np, self.constant)
+        # TODO: It would be possible to add them by merging tlist.
```

Comment:
It's a very small improvement, only specific case would use non matching `tlist` and computing the coefficient should be negligible compared to the matrix multiplication that follow. I guess it's a 1~2% improvement at best.
But since I expect it to be relatively easy to do, it feel lazy not to implement it...
I am keeping it the note, but not the to-do so it don't get caught in codeclimat etc.